File: src/2025-03-04-03-28-40--deepseek-chat-0304/python/exercises/practice/react/react.py
class InputCell:
    def __init__(self, initial_value):
        self._value = initial_value
        self._dependents = []

    @property
    def value(self):
        return self._value

    @value.setter
    def value(self, new_value):
        if self._value != new_value:
            self._value = new_value
            for cell in self._dependents:
                cell._update()

    def _add_dependent(self, cell):
        self._dependents.append(cell)


class ComputeCell:
    def __init__(self, inputs, compute_function):
        self._inputs = inputs
        self._compute = compute_function
        self._callbacks = set()
        self._dependents = []
        self._value = None
        
        # Register as dependent of all inputs
        for input_cell in inputs:
            input_cell._add_dependent(self)
            
        # Compute initial value
        self._update()

    @property
    def value(self):
        return self._value

    def _update(self):
        # Get current input values
        input_values = [input_cell.value for input_cell in self._inputs]
        # Compute new value
        new_value = self._compute(input_values)
        
        # Only update and notify if value changed
        if new_value != self._value:
            old_value = self._value
            self._value = new_value
            # Notify callbacks
            for callback in self._callbacks:
                callback(new_value)
            # Propagate to dependents
            for dependent in self._dependents:
                dependent._update()

    def add_callback(self, callback):
        self._callbacks.add(callback)

    def remove_callback(self, callback):
        self._callbacks.discard(callback)

    def _add_dependent(self, cell):
        self._dependents.append(cell)
    

File: src/2025-03-04-03-28-40--deepseek-chat-0304/javascript/exercises/practice/react/react.js
export class InputCell {
  constructor(value) {
    this.value = value;
    this.dependents = new Set();
  }

  setValue(value) {
    if (this.value !== value) {
      this.value = value;
      this.notifyDependents();
    }
  }

  addDependent(cell) {
    this.dependents.add(cell);
  }

  notifyDependents() {
    for (const cell of this.dependents) {
      cell.update();
    }
  }
}

export class ComputeCell {
  constructor(inputCells, fn) {
    this.inputCells = inputCells;
    this.computeFn = fn;
    this.callbacks = new Set();
    this.value = this.compute();
    
    for (const inputCell of inputCells) {
      if (inputCell instanceof InputCell) {
        inputCell.addDependent(this);
      }
    }
  }

  compute() {
    const values = this.inputCells.map(cell => cell.value);
    return this.computeFn(values);
  }

  update() {
    const newValue = this.compute();
    if (newValue !== this.value) {
      const oldValue = this.value;
      this.value = newValue;
      this.notifyCallbacks(oldValue);
    }
  }

  addCallback(cb) {
    this.callbacks.add(cb);
  }

  removeCallback(cb) {
    this.callbacks.delete(cb);
  }

  notifyCallbacks(oldValue) {
    for (const cb of this.callbacks) {
      cb(this, oldValue);
    }
  }
}

export class CallbackCell {
  constructor(fn) {
    this.fn = fn;
    this.values = [];
  }

  call(cell, oldValue) {
    this.values.push(this.fn(cell, oldValue));
  }
}
